### PR TITLE
Refactored with Statically (compile-time) limited Dynamic payload buffer. 

### PR DESCRIPTION
**Pull request by RichardKroesen**
### Brief
This is the initial concept version of the CayenneLPP library which is optimized for LoRaWAN-IoT lab requirements. 
It is setup in a way that the global assignment conditions are met (Open-closed principle applied). Only additional implementation is missing.

The class requires a at compile time a MaxSize value, passed as a template value. As a constructor parameter the operational/allocated size, this is the what the object actually allocates dynamically as a uint8_t array. The array is dynamically allocated using smart pointers (std::unique_prt). 

### Important Refactoring Changes
The library changed to a **header-only**, due to consequence of template based decision.

Use of smart pointers and modern C++ practices, the C++ standard is atleast ISO C++ 17. Memory leaks and huge responsibility of manual memory management is prevented in this way. 

### To be Realised List: 
For the following functions the prototypes are already created, just need the logic: 
- [ ] Add function for: bit (byte) & byte (one function);
- [ ] Add function: uint16_t words
- [ ] Add function: uint32_t words
- [ ] Add function: float words 

### Missing Functionality
The following features are not yet prepared: 
- Packet version number or type; 

### Code Example
**Creating a payload buffer**: `CayenneLPP<100> buf(15);` 
The allocated size of the buffer is 15. When a next buffer must be created it cannot go above 100. 
So, `CayenneLPP<100> buf(100);` is possible at run-time. But `CayenneLPP<100> buf(101);` Not.

### What to Do:
Here some tips and guide line of what to do in case of doubt:
1. Check the code changes on the right side.
2. Leave comments as feedback for me on the code lines or at the PR itself. 
3. Check the review approved when you are ok with the implementation.
4. Please DO NOT: merge or make commits this up to the PR-caller.

### CODE REVIEW

File: src/CayenneLPP.hpp [hunk 1]
```diff
@@ -1,41 +1,139 @@
...
+    /* END of Additional Functionality */
+
+    // To be Reviewed:
+    // RK: Just fancy functions of add_wordX...
+    uint8_t add_digital_input(uint8_t channel, uint8_t value)
+    {
+        if (currentIndex + LPP_DIGITAL_INPUT_SIZE > operationalSize)
+        {
+            // Handle error
+            return 1;
```
Comment (TristanBosveld) [inline src/CayenneLPP.hpp]: In this function you return a '1' for the error, while in the other functions you return a '0'. I would make these value's the same

File: src/CayenneLPP.hpp [hunk 2]
```diff
@@ -1,41 +1,139 @@
...
+        }
+        buffer[currentIndex++] = channel;
+        buffer[currentIndex++] = LPP_DIGITAL_INPUT;
+        buffer[currentIndex++] = value;
+        return 0; // Success
```
Comment (TristanBosveld) [inline src/CayenneLPP.hpp]: Don't you also want to return curent index in this function? Like in the other functions?

**Review by TristanBosveld — APPROVED** [review]

File: src/CayenneLPP.hpp (hunk 1 above)
Comment (RichardKroesen) [reply src/CayenneLPP.hpp]: Jup you are right, rookie mistake from me haha

File: src/CayenneLPP.hpp (hunk 2 above)
Comment (RichardKroesen) [reply src/CayenneLPP.hpp]: Indeed, fixed with the latest commit.

**Pull request merged**